File: FirmwareSource/libraries/CmdMessenger/extras/VisualBasic/TemperatureControl/LoggingView.vb
Imports Microsoft.VisualBasic
Imports System.Drawing
Imports System.Windows.Forms

Namespace Tools
	Partial Public Class LoggingView
		Inherits ListBox
		Public Sub New()
			InitializeComponent()
						SetStyle(ControlStyles.OptimizedDoubleBuffer Or ControlStyles.ResizeRedraw Or ControlStyles.UserPaint, True)
			DrawMode = DrawMode.OwnerDrawFixed
			FollowLastItem = True
			MaxEntriesInListBox = 3000
		End Sub

		Private privateFollowLastItem As Boolean
		Public Property FollowLastItem() As Boolean
			Get
				Return privateFollowLastItem
			End Get
			Set(ByVal value As Boolean)
				privateFollowLastItem = value
			End Set
		End Property
		Private privateMaxEntriesInListBox As Integer
		Public Property MaxEntriesInListBox() As Integer
			Get
				Return privateMaxEntriesInListBox
			End Get
			Set(ByVal value As Integer)
				privateMaxEntriesInListBox = value
			End Set
		End Property


		Public Sub AddEntry(ByVal item As Object)
			BeginUpdate()
			Items.Add(item)

			If Items.Count > MaxEntriesInListBox Then
				Items.RemoveAt(0)
			End If

			If FollowLastItem Then
				TopIndex = Items.Count - 1
			End If
			EndUpdate()
		End Sub

		Protected Overrides Sub OnDrawItem(ByVal e As DrawItemEventArgs)
			If Items.Count > 0 Then
				e.DrawBackground()
				e.Graphics.DrawString(Items(e.Index).ToString(), e.Font, New SolidBrush(ForeColor), New PointF(e.Bounds.X, e.Bounds.Y))
			End If
			MyBase.OnDrawItem(e)
		End Sub
		Protected Overrides Sub OnPaint(ByVal e As PaintEventArgs)
			Dim iRegion = New Region(e.ClipRectangle)
			e.Graphics.FillRegion(New SolidBrush(BackColor), iRegion)
			If Items.Count > 0 Then
				For i As Integer = 0 To Items.Count - 1
					Dim irect = GetItemRectangle(i)
					If e.ClipRectangle.IntersectsWith(irect) Then
						If (SelectionMode = SelectionMode.One AndAlso SelectedIndex = i) OrElse (SelectionMode = SelectionMode.MultiSimple AndAlso SelectedIndices.Contains(i)) OrElse (SelectionMode = SelectionMode.MultiExtended AndAlso SelectedIndices.Contains(i)) Then
							OnDrawItem(New DrawItemEventArgs(e.Graphics, Font, irect, i, DrawItemState.Selected, ForeColor, BackColor))
						Else
							OnDrawItem(New DrawItemEventArgs(e.Graphics, Font, irect, i, DrawItemState.Default, ForeColor, BackColor))
						End If
						iRegion.Complement(irect)
					End If
				Next i
			End If
			MyBase.OnPaint(e)
		End Sub
	End Class
End Namespace

File: FirmwareSource/lib/CmdMessenger/extras/CSharp/TemperatureControl/LoggingView.cs
using System.Drawing;
using System.Windows.Forms;

namespace Tools
{
    public partial class LoggingView : ListBox
    {
        public LoggingView()
        {
            InitializeComponent();
                        SetStyle(
                ControlStyles.OptimizedDoubleBuffer |
                ControlStyles.ResizeRedraw |
                ControlStyles.UserPaint,
                true);
            DrawMode = DrawMode.OwnerDrawFixed;
            FollowLastItem = true;
            MaxEntriesInListBox = 3000;
        }

        public bool FollowLastItem{ get; set; }
        public int MaxEntriesInListBox { get; set; }


        public void AddEntry(object item)
        {
            BeginUpdate();
            Items.Add(item);

            if (Items.Count > MaxEntriesInListBox)
            {
                Items.RemoveAt(0);
            }

            if (FollowLastItem) TopIndex = Items.Count - 1;
            EndUpdate();
        }

        protected override void OnDrawItem(DrawItemEventArgs e)
        {
            if (Items.Count > 0)
            {
                e.DrawBackground();
                e.Graphics.DrawString(Items[e.Index].ToString(), e.Font, new SolidBrush(ForeColor), new PointF(e.Bounds.X, e.Bounds.Y));
            }
            base.OnDrawItem(e);
        }
        protected override void OnPaint(PaintEventArgs e)
        {
            var iRegion = new Region(e.ClipRectangle);
            e.Graphics.FillRegion(new SolidBrush(BackColor), iRegion);
            if (Items.Count > 0)
            {
                for (int i = 0; i < Items.Count; ++i)
                {
                    var irect = GetItemRectangle(i);
                    if (e.ClipRectangle.IntersectsWith(irect))
                    {
                        if ((SelectionMode == SelectionMode.One && SelectedIndex == i)
                        || (SelectionMode == SelectionMode.MultiSimple && SelectedIndices.Contains(i))
                        || (SelectionMode == SelectionMode.MultiExtended && SelectedIndices.Contains(i)))
                        {
                            OnDrawItem(new DrawItemEventArgs(e.Graphics, Font,
                                irect, i,
                                DrawItemState.Selected, ForeColor,
                                BackColor));
                        }
                        else
                        {
                            OnDrawItem(new DrawItemEventArgs(e.Graphics, Font,
                                irect, i,
                                DrawItemState.Default, ForeColor,
                                BackColor));
                        }
                        iRegion.Complement(irect);
                    }
                }
            }
            base.OnPaint(e);
        }
    }
}
